File: android-agent/app/src/main/kotlin/com/remoteagent/cryptokit/Ckx1Aead.kt
package com.remoteagent.cryptokit

import javax.crypto.Cipher
import javax.crypto.spec.IvParameterSpec
import javax.crypto.spec.SecretKeySpec

/**
 * ChaCha20-Poly1305 AEAD only. Nonce must be 12 bytes; key 32 bytes.
 */
object Ckx1Aead {
    init {
        BcProvider.ensureInstalled()
    }

    fun encrypt(key: ByteArray, nonce: ByteArray, plaintext: ByteArray, aad: ByteArray): ByteArray {
        require(key.size == Ckx1Algorithm.KEY_SIZE) { "key must be 32 bytes" }
        require(nonce.size == Ckx1Algorithm.NONCE_SIZE) { "nonce must be 12 bytes" }
        val cipher = Cipher.getInstance("ChaCha20-Poly1305", "BC")
        cipher.init(Cipher.ENCRYPT_MODE, SecretKeySpec(key, "ChaCha20"), IvParameterSpec(nonce))
        if (aad.isNotEmpty()) cipher.updateAAD(aad)
        return cipher.doFinal(plaintext)
    }

    fun decrypt(key: ByteArray, nonce: ByteArray, ciphertext: ByteArray, aad: ByteArray): ByteArray {
        require(key.size == Ckx1Algorithm.KEY_SIZE) { "key must be 32 bytes" }
        require(nonce.size == Ckx1Algorithm.NONCE_SIZE) { "nonce must be 12 bytes" }
        require(ciphertext.size >= Ckx1Algorithm.TAG_SIZE) { "ciphertext too short" }
        val cipher = Cipher.getInstance("ChaCha20-Poly1305", "BC")
        cipher.init(Cipher.DECRYPT_MODE, SecretKeySpec(key, "ChaCha20"), IvParameterSpec(nonce))
        if (aad.isNotEmpty()) cipher.updateAAD(aad)
        return cipher.doFinal(ciphertext)
    }
}

File: server/internal/cryptokit/ckx1_aead.go
package cryptokit

import (
	"errors"

	"golang.org/x/crypto/chacha20poly1305"
)

// Ckx1AEADEncrypt seals plaintext with ChaCha20-Poly1305 and AAD. Nonce must be 12 bytes.
func Ckx1AEADEncrypt(key, nonce, plaintext, aad []byte) ([]byte, error) {
	if len(key) != CKX1KeySize {
		return nil, errors.New("key must be 32 bytes")
	}
	if len(nonce) != chacha20poly1305.NonceSize {
		return nil, errors.New("nonce must be 12 bytes")
	}
	aead, err := chacha20poly1305.New(key)
	if err != nil {
		return nil, err
	}
	return aead.Seal(nil, nonce, plaintext, aad), nil
}

// Ckx1AEADDecrypt opens ciphertext with ChaCha20-Poly1305 and AAD.
func Ckx1AEADDecrypt(key, nonce, ciphertext, aad []byte) ([]byte, error) {
	if len(key) != CKX1KeySize {
		return nil, errors.New("key must be 32 bytes")
	}
	if len(nonce) != chacha20poly1305.NonceSize {
		return nil, errors.New("nonce must be 12 bytes")
	}
	aead, err := chacha20poly1305.New(key)
	if err != nil {
		return nil, err
	}
	return aead.Open(nil, nonce, ciphertext, aad)
}
